<html>
<head>
  <title>CSC104 - Lab 3 - Strands of DNA</title>
  <link rel="stylesheet" type="text/css" title="Default" href="../style.css"> 
</head>

<body>
<h1><a href="../index.html">CSC104</a> - Lab 3<br>Strands of DNA</h1>
<h3>Assigned September 12th 2 p.m.<br>
Due September 17th 2 p.m.</h3>
<hr>

<h2>Overview</h2>
DNA molecules provide the basis for all known life on Earth. In this lab we will explore the
String data type by analyzing and manipulating strands of DNA.  I first provide a brief
overview of DNA and RNA, followed by steps in Python to carry out these molecular processes.

<h2>Materials</h2>

<ul>
	<li><a href="http://docs.python.org/lib/node42.html">String Functions</a> in Python
	<li><a href="../data/histone_h4.txt">histone_h4.txt</a>
	<li>Lab Partner
</ul>

<h2>Description</h2>
<img class="entryimageleft" src="http://upload.wikimedia.org/wikipedia/en/f/f0/DNA_Overview.png">
The search for intelligent life in the universe, according to some conservative
estimates for the Drake equation, may take a very long time before we are successful.
It is much more likely that we will first find evidence of life at a much more basic level.
<a href="http://astrobiology.arc.nasa.gov/">Astrobiology</a> is the study of life itself
and where it might exist in the universe, intelligent or not.  By studying life on 
Earth and the conditions under which it survives, be they normal temperate climates
or harsh extremes where only <a href="http://en.wikipedia.org/wiki/Extremophile">
Extremophiles</a> can be found, we can begin to reason about other locations
in our solar system and galaxy that might also harbor life.  
<p>
One thread common to all life on Earth is Deoxyribonucleic acid (DNA).
DNA is composed of 
long strands of four basic molecules: Adenine, Cytosine, Guanine and Thymine. These four bases coil 
together to form a double-helix structure, and we have 23 such structures called chromosomes in
every cell of our body. The bases have the interesting property of pairing up across the 
double-helix, such that Adenine always matches Thymine, and Cytosine always matches Guanine.
For our purposes as computer scientists, we can abbreviate these bases as A, C, G, and T.  
<p>
DNA contains segments called genes detailing how every piece 
of our cells should work. 
A gene becomes a protein through a process known as the
<a href="http://en.wikipedia.org/wiki/Central_dogma_of_molecular_biology">Central dogma
of molecular biology</a>: DNA is transcribed into RNA (Ribonucleic Acid) and then translated
into proteins. Proteins are the machinery that makes our cells work; they form the cell walls, 
transport molecules around the cell, and in general make everything work.  It is this process
of DNA to RNA to Proteins that we will study in this lab, using Python and String functions
to analyze small pieces of DNA.  The process may seem a bit complex, but with Python we can
make very fast progress.

<h3>Step 1</h3>
We'll start with a small example to get familiar with how to analyze these DNA strings
with Python.  Our sample DNA for these first experiments will be 
<p>
<pre>s = "catgcttcgcataacatgactgct"
</pre>
<p>
Depending on the database we use to find our DNA sequences, some may be in lowercase, while others
are all capital letters.  To make our life easier, we can convert any string to all uppercase
by using the <code>string.upper</code>(<i>string</i>) function from the string module to 
eliminate any variations from the source.
<p>
<b>G-C Content</b>
<p>
After capitalization, we now have
<p>
<pre>s = "CATGCTTCGCATAACATGACTGCT"
</pre>
<p>One way to determine if there might be a gene
hiding in a strand of DNA is to examine it's G-C content.  Because of how genes are encoded, most 
genes have a balanced number of Gs and Cs in relation to the number of As and Ts.  To determine the 
G-C content of a DNA strand, we can add up the number of Gs and Cs we find, and divide this by 
the total number of bases in the string.  In this case, there would be 11 Gs and Cs out of 24 total 
bases, giving us 45.83%. Most genes have between 25% and 75% Gs and Cs, so this gives our strand
a good shot of being a gene.  Python lets us do this with the <code>string.count</code>(<i>string, 
substring</i>) function.
<p>
<b>Transcription from DNA to RNA</b>
<p>
Our next step is to transcribe the DNA into RNA.  RNA is a single-stranded molecule as opposed to
the double-stranded DNA.  This means it cannot replicate itself as does DNA, but RNA still 
plays many useful roles in cell function. Primarily, it is RNA that gets translated into 
proteins, so having these steps be separate eases the burden on DNA to process all the work. 
In the transcription
from DNA to RNA, every Thymine molecule is replace with Uracil.  So when manipulating
these strings in Python, we can replace T with U using the 
<code>string.replace</code>(<i>string, find, replace</i>) function.  This gives us the RNA strand
<p>
<pre>r = "CAUGCUUCGCAUAACAUGACUGCU"
</pre>
<p>
<b>Translation from RNA to proteins</b>
<p>
<img class="entryimageright" src="http://upload.wikimedia.org/wikipedia/en/d/d4/RNA-codons.png">
With our strand now in RNA form, we are ready to make the final step to proteins.  This works
by reading the <a href="http://en.wikipedia.org/wiki/Genetic_code">Genetic Code</a> of RNA. 
Just like words are composed of individual letters, genes are divided up into segments called
<b>codons</b>.  A codon consists of three sequential bases, so with 4 choices for each base 
(from A, C, G, and U), this gives us 4 X 4 X 4 = 64 different combinations.
Each of these 64 possibilities translates into one of 20 
<a href="http://en.wikipedia.org/wiki/List_of_standard_amino_acids">amino acids</a>
(redundancy is built into this system to decrease the effects of random mutations),
and it is the conjunction of these amino acids that ultimately fold up and create proteins.
Not all of the bases in a strand of RNA will encode for a gene, and actually, only 10% - 20% of
our current DNA is known to encode genes; the other 80% - 90% of our genome is mostly 
unexplained and unexplored for different modes of functionality. 
<p>
When you are reading a strand of RNA, there are two important locations to find: where to 
<b>start</b>
reading and where to <b>stop</b> reading.  
There are actually three different ways to change a strand into amino acids.  
We could start with the first base at index 0,
such that our codons are substrings from 0:3, 3:6, 6:9, etc.  Staring with index 1
makes our substrings 1:4, 4:7, 7:10, etc, and starting with index 2 gives us 2:5, 5:8, 8:11, etc.
<p>
These initial indices (0, 1, 2) define the <b>reading frame</b> for this gene.  
In English we start every sentence with a capital letter, and stop each sentence with a period.
RNA uses an interesting way of signaling the start of a gene by designating one particular codon,
AUG, as the start codon.  For our purposes, wherever you find AUG, this marks the start of a gene;
this also defines our reading frame, which can be found by modding the initial index of the start
codon by 3.  In our sample sequence above, we can use the function 
<code>string.find</code>(<i>string, substring</i>) to locate the index for "AUG", and determine
that the reading frame is 1 % 3 = 1.
<p>
For robustness, there are three codons which tell us when to stop translating RNA into 
amino acids: UAG, UGA and UAA. As soon as any one of these three stop codons is found <b>
in the same reading frame after the start codon</b>, translation stops and the protein is complete.
In our RNA strand, we first find UAA at index 11, however 11 % 3 = 2 and is not in the same
reading frame as the start codon.  Our next instance of a stop codon is at index 16, and since
16 % 3 = 1, we have found our stop codon.  This means that the gene to be translated is
<p>
<pre>gene = "CUUCGCAUAACA"
</pre>
<p>
We will leave the final translation step into the actual amino acids for a later 
project.  For now, it will suffice to find the start and stop codons within our RNA 
strand.
<p>

<h3>Step 2</h3>
Now that you have some familiarity with the processing of DNA, we will be 
using the interpreter to analyze very long DNA strings.
<p>
Download <a href="../data/histone_h4.txt"><code>histone_h4.txt</code></a> to your
working Python directory (the default where your files are saved).  
This relatively small section of DNA (most genes are between 1000 and 10,000 bases long)
that encodes for 
<a href="http://en.wikipedia.org/wiki/Histone">Histone H4</a>.
Histone H4 plays a major role in how the DNA strands wind up to form chromosomes,
and this gene is highly conserved across almost all 
<a href="http://en.wikipedia.org/wiki/Eukaryote">Eukaryotic organisms</a>
(those which have a nucleus).
<p>
Perform the following steps on the DNA strand, and record the results in 
<code>lab3_evaluation.txt</code>:
<ol>
<li>Open this file through the interpreter and read in the data into a string.
<li>Convert this string to uppercase and record
the first 10 bases.
<li>Find and record the GC 
content of the DNA as a percent of the overall length of the DNA.  
<li>Translate this gene into RNA. Record how many bases were changed from T to U.
<li>Determine the reading frame for the first gene present in the RNA.
<li>Find the segment of RNA that encodes the first gene, excluding the start and stop codon.
This will take some trial and error from you; use the functions described in the string module.
Remember that the stop codon must be in the same reading frame as the start codon.
Record the locations of both the start and stop codon.
</ol>
<p>
<h2>What to Hand In</h2>
Log in to <code>cs.centenary.edu</code> through either Secure FTP or WinSCP using your
cs login and password.  Create a subdirectory from <code>csc104</code> 
called <code>lab3</code> and make it read, write and execute privileges for only the
owner.  Write up your results from step 2 into a file called 
<code>lab3_evaluation.txt</code> and copy it into this directory. 
<p>
You must hand in:
<ul>
 	<li>lab3_evaluation.txt
</ul>
<hr>
<small>&copy; Mark Goadrich 2007, Centenary College of Louisiana</small>

</body>